libraries/AP_Baro/AP_Baro_ICM20789.cpp

```diff
@@ -0,0 +1,318 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/I2CDevice.h>
+#include <utility>
+
+#include <AP_Common/AP_Common.h>
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Math/AP_Math.h>
+#include <AP_BoardConfig/AP_BoardConfig.h>
+#include "AP_Baro_ICM20789.h"
+
+#include <utility>
+#include <stdio.h>
+
+#include <AP_Math/AP_Math.h>
+#include <DataFlash/DataFlash.h>
+
+extern const AP_HAL::HAL &hal;
+
+/*
+  CMD_READ options. The draft datasheet doesn't specify, but it seems
+  Mode_1 has a conversion interval of 2ms. Mode_3 has a conversion
+  interval of 20ms. Both seem to produce equally as smooth results, so
+  presumably Mode_3 is doing internal averaging
+ */
+#define CMD_READ_PT_MODE_1 0x401A
+#define CMD_READ_PT_MODE_3 0x5059
+#define CMD_READ_TP_MODE_1 0x609C
+#define CMD_READ_TP_MODE_3 0x70DF
+
+#define CONVERSION_INTERVAL_MODE_1 2000
+#define CONVERSION_INTERVAL_MODE_3 20000
+
+// setup for Mode_3
+#define CMD_READ_PT CMD_READ_PT_MODE_3
+#define CONVERSION_INTERVAL CONVERSION_INTERVAL_MODE_3
+
+#define CMD_SOFT_RESET     0x805D
+#define CMD_READ_ID        0xEFC8
+
+#define debug(fmt, args...)   hal.console->printf(fmt, ##args)
+
+/*
+  constructor
+ */
+AP_Baro_ICM20789::AP_Baro_ICM20789(AP_Baro &baro, AP_HAL::OwnPtr<AP_HAL::Device> _dev)
+    : AP_Baro_Backend(baro)
+    , dev(std::move(_dev))
+{
+}
+
+AP_Baro_Backend *AP_Baro_ICM20789::probe(AP_Baro &baro,
+                                         AP_HAL::OwnPtr<AP_HAL::Device> dev)
+{
+    debug("Probing for ICM20789 baro\n");
+    if (!dev) {
+        return nullptr;
+    }
+    AP_Baro_ICM20789 *sensor = new AP_Baro_ICM20789(baro, std::move(dev));
+    if (!sensor || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+    return sensor;
+}
+
+
+bool AP_Baro_ICM20789::spi_init(void)
+{
+    dev_icm->set_read_flag(0x80);
+
+    dev_icm->set_speed(AP_HAL::Device::SPEED_LOW);
+    uint8_t whoami = 0;
+    uint8_t v;
+        
+    dev_icm->read_registers(0x6A, &v, 1);
+    dev_icm->write_register(0x6B, 0x01);
+    
+    hal.scheduler->delay(1);
+    dev_icm->write_register(0x6A, 0x10);
```

Comment:
You are setting an IMU register without reading whatever the INS set there... This will give us trouble if later we need to set something in this register there.  It would be nice to keep this IMU part on the INS driver like it is when we have a mag rather than a baro.

And... now that we have access to the plain I2C/SPI dev, we can also simplify the AuxBus infra a little bit.

Reply:
@lucasdemarchi Should I move the initialisation into the IMU, or is it OK where it is for the time being?

Reply:
baro is initialised before ins, which is why we need to init here, as ins doesn't yet exist

Reply:
long ago I had this same problem when MS5611 was on the auxiliary bus of MPU6K... hence the infra for that.   So if later on the INS drivers writes to this register, whatever we did here is lost.  I think it's much safer to "keep INS registers on the INS driver and baro registers on the baro driver".